### DISCUSSION TITLE
IDEA.Statica.Plugin API - ConnectionHiddenCheckClient - Setting analysis type?

### DISCUSSION THREAD

**Discussion by mitchell-tesch** [idea-statica/ideastatica-public#295] [Help Q&A] [answered]
Hi,

Is it possible to set the analysis type of each item in the connection file, using this API?

Kind regards,

Mitch

**Comment by nathanlukeIDEA** [comment]
Hi Mitchell, 

It is not currently possible to set the solver mode through the API. 

What solver mode are you most trying to set? 

Cheers
Nathan

**Reply by mitchell-tesch** [reply]
Hi Nathan,

Hope you are well.

In the current scenario we are trying to batch a multiple of existing connection files for a Design Resistance (DR) analysis. These connections were previous solved using a stress/strain with buckling (EPS) analysis. 

Generally we find EPS and DR would be the most common analysis types we are trying to set. To date we have been setting by opening each file changing the analysis type, saving and closing. I have also noticed that when we have a file set to EPS and calculate through the Plugin API we do not get any buckling results, is it possible to also enable buckling in an EPS analysis via the API.

On occasions I am also finding that when we solve a model performing a DR analysis initiated through the API that no results are returned (empty results JSON). I am unsure if this is due to the size of the model (number of load cases) - I suspect the solver is crashing? Although when initiated through the GUI the solver does complete in full (it does take a decent amount of time ~1.5hrs). I have attached an example of a connection file to which this this occurs. It would be good to get your thoughts? 

If it helps I can also share my code.

Kind regards,

Mitch
[C09-2055.zip](https://github.com/idea-statica/ideastatica-public/files/11954880/C09-2055.zip)

**Reply by mitchell-tesch** [reply]
Further to the above I tested on a smaller version of the same model. i.e. same model with only three load cases; load case 2, 22 and 23.
This completes and returns results, so I do suspect the problem arises only on larger models. I have attached the smaller model for reference. 
[C09-2055_small.zip](https://github.com/idea-statica/ideastatica-public/files/11955643/C09-2055_small.zip)

**Reply by nathanlukeIDEA** [reply, marked as answer]
Hi Mitchell, 

In general, the API is not currently set up to effectively output results for analysis types other than EPS. We will need to improve the API a little to be able to output the results of an DR calculation in a nicer manner. Some answers to your queries below.

> Generally we find EPS and DR would be the most common analysis types we are trying to set. To date we have been setting by opening each file changing the analysis type, saving and closing. I have also noticed that when we have a file set to EPS and calculate through the Plugin API we do not get any buckling results, is it possible to also enable buckling in an EPS analysis via the API.

We are in the process of adding the ability to run calculations with buckling, but as it currently stands buckling will not be run with the calculation when triggered through the API.

> On occasions I am also finding that when we solve a model performing a DR analysis initiated through the API that no results are returned (empty results JSON). I am unsure if this is due to the size of the model (number of load cases) - I suspect the solver is crashing? Although when initiated through the GUI the solver does complete in full (it does take a decent amount of time ~1.5hrs). I have attached an example of a connection file to which this this occurs. It would be good to get your thoughts?

I will add this as an issue. We will need to do some more testing on this. I am not sure whether internally an issue is arising due to the shear amount of data we are trying to provide through the calculation result method.

**Reply by mitchell-tesch** [reply]
Thank you @nathanlukeIDEA, appreciate your help.

**Comment by nathanlukeIDEA** [comment]
Hi Mitchell, 

Can you send me a snip of your code? I want to double-check which methods you use for calculation and results extraction.

Cheers
Nathan

**Reply by mitchell-tesch** [reply]
Hi Nathan,

Yes, no problems, the exact code I am using can be found here:
https://github.com/mitchell-tesch/PyDEA-StatiCa/blob/main/examples/IdeaStatiCaDesignResistanceBatch/IdeaStatiCaConnDesignResistanceBatch.py

Sorry about it being a file in a larger repo, I am using a Python wrapper around the IDEAStatica.Plugin API.

Kind regards,

Mitch

**Reply by umesh14398** [reply]
Hi @mitchell-tesch 
Good Morning,

Hope you are doing well. 
This is Uma Mahesh from India. I am reaching out to you for help regarding Idea Statica. I am working on an automation task in Idea Statica. I am trying to a establish connection with IdeaStatica using Python but I am not able to execute it. Please find the code snippet below. I am facing a SessionTimeout Exception whenever i run the script. I am not really sure what could be the problem. Can you please help me out in this regard? Looking forward hearing from you.
![image](https://github.com/idea-statica/ideastatica-public/assets/43907961/87efc75d-4065-483a-bee6-22b40282dec5)

**Reply by mitchell-tesch** [reply]
Hi @umesh14398 ,

Hope you are well.

Your code appears to be valid. The session System.TimeoutException you are seeing usually indicates the inability for the IDEA StatiCa process to establish connection to a valid license. I am able to replicate on my end with the below snippet. If I am unable to obtain a valid license I will see the System.TimeoutException, otherwise I will see the return of a valid client with process id.

```python
import clr
import sys

idea_path = r"C:\Program Files\IDEA StatiCa\StatiCa 23.0"
sys.path.append(idea_path)

clr.AddReference("IdeaStatiCa.Plugin")

import IdeaStatiCa.Plugin as isapi

factory = isapi.ConnHiddenClientFactory(idea_path)
client: isapi.ConnectionHiddenCheckClient = factory.Create()

print(f"Service: {client.Service}\n",
      f"CalculatorID: {client.HiddenCalculatorId}\n",
      f"Channel: {client.Channel}")

factory.Dispose()
```

Are you able to open the `License.exe` within the IDEA StatiCa path and Reserve a valid Steel enhanced license and try your code again?

PS. if it is helpful I have implemented most of the key functionality of the IDEAStatiCa.Plugin API in a Python library: [PyDEA StatiCa](https://github.com/mitchell-tesch/PyDEA-StatiCa/tree/main).

**Reply by umesh14398** [reply]
Hi @mitchell-tesch 
I ran License.exe and checked the available products. Please find the code snippet below. Do I need to reserve the license to make the connection with Idea Statica?
![image](https://github.com/idea-statica/ideastatica-public/assets/43907961/afe68532-07fd-46ad-8a31-5530467c831f)

**Reply by mitchell-tesch** [reply]
@umesh14398  - technically you do not need to Reserve the license, but you do need to have one accessible and available and the time of initialising the HiddenClient To avoid any potential issues of licenses becoming unavailable during execution I would recommend Reserving before executing.

From your screen shot above I notice you have a "Steel" license. I am unsure on the difference between a "Steel" license and a "Steel Enhanced" license as I only have "Steel Enhanced". @nathanlukeIDEA  could this cause any issues with Plugin API access?

**Reply by umesh14398** [reply]
Hi @mitchell-tesch 
@nathanlukeIDEA  is on vacation and he will be back on 6th Nov. He told me that there might be a bug in the plug-in and asked me to update the Idea Statica application to the latest version and keep trying with the same script. I didn't find any luck.

**Reply by mitchell-tesch** [reply]
> Hi @mitchell-tesch 
> 
> @nathanlukeIDEA  is on vacation and he will be back on 6th Nov. He told me that there might be a bug in the plug-in and asked me to update the Idea Statica application to the latest version and keep trying with the same script. I didn't find any luck.

Okay, which version are you currently running? And after reserving a license do you see the exact same trace?

**Reply by umesh14398** [reply]
I am currently having Idea Static ver23.0.5, and I ran the script after reserving the license but still i could see the same issue.

**Reply by mpospisil** [reply]
see my comment https://github.com/idea-statica/ideastatica-public/discussions/212

**Reply by mpospisil** [reply]
you can download [ 23.1 beta  version](https://download.ideastatica.com/beta/idea-statica.msi) it should fix the issue with license

**Reply by umesh14398** [reply]
Hi @mpospisil 
Thank you, Martin. I will try with this beta version and see. Hope it resolves the issue.
